File: Client/Handler/Draw/Dx/DxBase.cs
using RAGE.Game;
using System;
using System.Collections.Generic;
using TDS_Client.Data.Enums;

namespace TDS_Client.Handler.Draw.Dx
{
    // UIResText.Draw is top aligned Graphics.DrawRect is center aligned (X & Y)
    public abstract class DxBase : IDisposable
    {
        public int FrontPriority;

        protected readonly List<DxBase> Children = new List<DxBase>();
        protected readonly DxHandler DxHandler;

        private bool _disposed = false;

        protected DxBase(DxHandler dxHandler, int frontPriority = 0, bool activated = true)
        {
            DxHandler = dxHandler;

            Activated = activated;
            FrontPriority = frontPriority;
            dxHandler.Add(this);
        }

        ~DxBase() => Dispose(false);

        public bool Activated { get; set; }

        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        public virtual void Draw()
        {
        }

        public virtual DxType GetDxType()
        {
            return DxType.None;
        }

        public virtual void Remove()
        {
            DxHandler.Remove(this);
            Dispose(true);
        }

        protected static int GetBlendValueInt(int currenttick, int start, int end, int starttick, int endtick)
        {
            float progress = (currenttick - starttick) / (endtick - starttick);
            if (progress > 1)
                progress = 1;
            return (int)Math.Floor(start + progress * (end - start));
        }

        protected static float GetBlendValue(float currentTick, float start, float end, float startTick, float endTick)
        {
            float progress = (currentTick - startTick) / (endTick - startTick);
            if (progress > 1)
                progress = 1;
            return start + progress * (end - start);
        }

        // To detect redundant calls
        protected virtual void Dispose(bool disposing)
        {
            if (_disposed)
                return;
            _disposed = true;

            if (disposing)
            {
                foreach (var child in Children)
                    if (!child._disposed)
                        child.Remove();
                Children.Clear();
            }
        }

        protected virtual int GetAbsoluteX(float x, bool relative, bool isText = false)
        {
            return (int)Math.Round(relative ? x * (isText ? 1920 : DxHandler.ResX) : x * (isText ? (1920 / DxHandler.ResX) : 1));
        }

        protected virtual int GetAbsoluteY(float y, bool relative, bool isText = false)
        {
            return (int)Math.Round(relative ? y * (isText ? 1080 : DxHandler.ResY) : y * (isText ? (1080 / DxHandler.ResY) : 1));
        }

        protected float GetRelativeX(float x, bool relative, bool isText = false)
        {
            return relative ? x : x / (isText ? 1920 : DxHandler.ResX);
        }

        protected float GetRelativeY(float y, bool relative, bool isText = false)
        {
            return relative ? y : y / (isText ? 1080 : DxHandler.ResY);
        }

        protected int GetTextAbsoluteHeight(float lineCount, float scale, Font font, bool relative)
        {
            int textHeight = GetAbsoluteY(Ui.GetTextScaleHeight(scale, (int)font), relative, true);

            // + 5 ... because of the margin between the lines
            textHeight = (int)(textHeight * lineCount + textHeight * 0.4 * (lineCount - 0.3));
            return textHeight;
        }
    }
}

File: Client/TypeScript/src/entities/draw/dx/dx-base.ts
import DxService from "../../../handler/draw/dx/dx.service";
import { getTextScaleHeight } from "natives";
import Font from "../../../data/enums/dx/font.enum";
import DxType from "../../../data/enums/dx/dx-type.enum";

abstract class DxBase {
    frontPriority: number;
    activated: boolean;

    protected readonly children: DxBase[] = [];
    protected readonly dxService: DxService;

    constructor(
        dxService: DxService, frontPriority: number = 0, activated: boolean = true) {

        this.dxService = dxService;

        this.activated = activated;
        this.frontPriority = frontPriority;

        this.dxService.add(this);
    }

    abstract draw(): void;
    abstract getDxType(): DxType;

    remove() {
        this.dxService.remove(this);

        for (const child of this.children) {
            child.remove();
        }
        this.children.length = 0;
    }

    protected getBlendValue(currentTick: number, start: number, end: number, startTick: number, endTick: number): number {
        const progress = Math.min(1, (currentTick - startTick) / (endTick - startTick));
        return Math.floor(start + progress * (end - start));
    }

    protected getAbsoluteX(x: number, relative: boolean, isText: boolean = false): number {
        return Math.round(relative ? x * (isText ? 1920 : this.dxService.resX) : x * (isText ? (1920 / this.dxService.resX) : 1));
    }

    protected getAbsoluteY(y: number, relative: boolean, isText: boolean = false): number {
        return Math.round(relative ? y * (isText ? 1080 : this.dxService.resY) : y * (isText ? (1080 / this.dxService.resY) : 1));
    }

    protected getRelativeX(x: number, relative: boolean, isText: boolean = false): number {
        return relative ? x : x / (isText ? 1920 : this.dxService.resX);
    }

    protected getRelativeY(y: number, relative: boolean, isText: boolean = false): number {
        return relative ? y : y / (isText ? 1080 : this.dxService.resY);
    }

    protected getTextAbsoluteHeight(lineCount: number, scale: number, font: Font, relative: boolean): number {
        let textHeight = this.getAbsoluteY(getTextScaleHeight(scale, font), relative, true);

        // + 5 ... because of the margin between the lines
        textHeight = Math.floor(textHeight * lineCount + textHeight * 0.4 * (lineCount - 0.3));
        return textHeight;
    }
}

export default DxBase;
